### PR TITLE
Corrects workspace project tree refresh behavior for adding new projects to the workspace

### PR DESCRIPTION
**Pull request by Benjin**
The PR fixes https://github.com/microsoft/azuredatastudio/issues/16227

Reordered the steps for determining whether a project is already in the workspace, relying on a cached copy of opened projects instead of globbing all workspace folders every check for scaling perf reasons.

Also promotes the refresh button in the viewbar title to be a primary action so that users can manually (outside of ADS) add new projects to a folder and easily get them detected (rather than using a bunch of recursive folder watchers)

### CODE REVIEW

File: extensions/data-workspace/src/services/workspaceService.ts [hunk 1]
```diff
@@ -47,46 +51,58 @@ export class WorkspaceService implements IWorkspaceService {
...
+		if (newProjectAdded) {
 			this._onDidWorkspaceProjectsChange.fire();
 		}
 
-		if (newWorkspaceFolders.length > 0) {
-			// Add to the end of the workspace folders to avoid a restart of the extension host if we can
-			vscode.workspace.updateWorkspaceFolders(vscode.workspace.workspaceFolders?.length || 0, undefined, ...(newWorkspaceFolders.map(folder => ({ uri: vscode.Uri.file(folder) }))));
+		// 4. Signal caller if not all projects were new
+		if (anyProjectsAlreadyOpen) {
+			throw new Error('Some projects were already open in the workspace.');
```
Comment (Benjin) [inline extensions/data-workspace/src/services/workspaceService.ts:101]: Localize

File: extensions/data-workspace/src/services/workspaceService.ts [hunk 2]
```diff
@@ -95,19 +111,21 @@ export class WorkspaceService implements IWorkspaceService {
 		return projectTypes;
 	}
 
-	async getProjectsInWorkspace(ext?: string): Promise<vscode.Uri[]> {
-		const projectPromises = vscode.workspace.workspaceFolders?.map(f => this.getAllProjectsInFolder(f.uri));
-		if (!projectPromises) {
-			return [];
+	public async getProjectsInWorkspace(ext?: string, refreshFromDisk: boolean = false): Promise<vscode.Uri[]> {
```
Comment (Benjin) [inline extensions/data-workspace/src/services/workspaceService.ts:114]: docstrings

File: extensions/data-workspace/src/services/workspaceService.ts [hunk 3]
```diff
@@ -47,46 +51,58 @@ export class WorkspaceService implements IWorkspaceService {
 		}
 	}
 
-	async addProjectsToWorkspace(projectFiles: vscode.Uri[]): Promise<void> {
-		if (!projectFiles || projectFiles.length === 0) {
-			return;
-		}
+	public async addProjectsToWorkspace(projectFiles: vscode.Uri[]): Promise<void> {
+		// 1. Update workspace folders if any of the new projects's locations aren't already included
```
Comment (Benjin) [inline extensions/data-workspace/src/services/workspaceService.ts:55]: "Include new workspace folders..." would be more clear

File: extensions/data-workspace/src/dataworkspace.d.ts [hunk 4]
```diff
@@ -18,7 +18,7 @@ declare module 'dataworkspace' {
 		 * Returns all the projects in the workspace
 		 * @param ext project extension to filter on. If this is passed in, this will only return projects with this file extension
 		 */
-		getProjectsInWorkspace(ext?: string): Promise<vscode.Uri[]>;
+		getProjectsInWorkspace(ext?: string, refreshFromDisk?: boolean): Promise<vscode.Uri[]>;
```
Comment (Benjin) [inline extensions/data-workspace/src/dataworkspace.d.ts:21]: update docstring with new param

File: extensions/data-workspace/src/common/interfaces.ts [hunk 5]
```diff
@@ -51,7 +51,7 @@ export interface IWorkspaceService {
 	/**
 	 * Gets the project files in current workspace
 	 */
-	getProjectsInWorkspace(): Promise<vscode.Uri[]>;
+	getProjectsInWorkspace(ext?: string, refreshFromDisk?: boolean): Promise<vscode.Uri[]>;
```
Comment (Charles-Gagnon) [inline extensions/data-workspace/src/common/interfaces.ts:54]: add params to doc comment

File: extensions/data-workspace/src/services/workspaceService.ts [hunk 6]
```diff
@@ -95,19 +111,21 @@ export class WorkspaceService implements IWorkspaceService {
...
 
-	async getProjectsInWorkspace(ext?: string): Promise<vscode.Uri[]> {
-		const projectPromises = vscode.workspace.workspaceFolders?.map(f => this.getAllProjectsInFolder(f.uri));
-		if (!projectPromises) {
-			return [];
+	public async getProjectsInWorkspace(ext?: string, refreshFromDisk: boolean = false): Promise<vscode.Uri[]> {
+		if (refreshFromDisk) {
+			const projectPromises = vscode.workspace.workspaceFolders?.map(f => this.getAllProjectsInFolder(f.uri));
+			if (!projectPromises) {
+				return [];
```
Comment (Charles-Gagnon) [inline extensions/data-workspace/src/services/workspaceService.ts:118]: You should set openedProjects to empty here 

(it'd probably be better to just remove the check and do `			const projectPromises = vscode.workspace.workspaceFolders?.map(f => this.getAllProjectsInFolder(f.uri)) ?? [];`

File: extensions/data-workspace/src/services/workspaceService.ts [hunk 7]
```diff
@@ -18,7 +18,11 @@ export class WorkspaceService implements IWorkspaceService {
 	private _onDidWorkspaceProjectsChange: vscode.EventEmitter<void> = new vscode.EventEmitter<void>();
 	readonly onDidWorkspaceProjectsChange: vscode.Event<void> = this._onDidWorkspaceProjectsChange?.event;
 
-	constructor() { }
+	private openedProjects: vscode.Uri[] = [];
+
+	constructor() {
+		this.getProjectsInWorkspace(undefined, true);
```
Comment (Charles-Gagnon) [inline extensions/data-workspace/src/services/workspaceService.ts:24]: You may want to consider having re-entry protection for this stuff, currently you could potentially have issues if this is called multiple times in a row.

File: extensions/data-workspace/src/main.ts [hunk 8]
```diff
@@ -20,8 +20,8 @@ export async function activate(context: vscode.ExtensionContext): Promise<IExten
 	const workspaceService = new WorkspaceService();
 
 	const workspaceTreeDataProvider = new WorkspaceTreeDataProvider(workspaceService);
-	context.subscriptions.push(vscode.workspace.onDidChangeWorkspaceFolders(() => {
-		workspaceTreeDataProvider.refresh();
+	context.subscriptions.push(vscode.workspace.onDidChangeWorkspaceFolders(async () => {
+		await workspaceTreeDataProvider.refresh();
```
Comment (udeeshagautam) [inline extensions/data-workspace/src/main.ts:24]: Will the await slow down UI?

**Comment by udeeshagautam** [comment]
Looks fine to me but seems big for this ask mode. Please test well before checkin :)

File: extensions/data-workspace/src/main.ts (hunk 8 above)
Comment (Charles-Gagnon) [reply extensions/data-workspace/src/main.ts:24]: This is already in a callback so adding this isn't changing anything

**Review by Charles-Gagnon — APPROVED** [review]

**Comment by Benjin** [comment]
> Looks fine to me but seems big for this ask mode. Please test well before checkin :)

All tests are passing, and I've tried all the workspace + project add/remove scenarios I can think of; everything seems to be working as expected.

**Pull request merged**